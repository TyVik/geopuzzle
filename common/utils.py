import string

from django.utils.crypto import get_random_string


def random_string():
    return get_random_string(12, string.ascii_lowercase + string.digits)
