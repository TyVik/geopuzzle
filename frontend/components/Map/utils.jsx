/* global google */


function decodePolygon(polygon) {
  return polygon.map(polygon => (google.maps.geometry.encoding.decodePath(polygon)));
}

export { decodePolygon };
