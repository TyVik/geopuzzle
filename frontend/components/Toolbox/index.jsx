'use strict';
/* global google */
import React from "react";
import {Panel} from "react-bootstrap";
import { Scrollbars } from 'react-custom-scrollbars';
import localization from "../../localization";
import "./index.css";


const NameListItem = (props) => {
    if (!props.polygon.draggable) {
        return <li
            className={"clickable list-group-item list-group-item-" + (props.polygon.isSolved ? 'success' : 'danger')}
            onClick={props.click}>
            {props.polygon.infobox.name}
        </li>;
    } else {
        return <li className="list-group-item list-group-item-danger">&nbsp;</li>;
    }
};


class Toolbox extends React.Component {
    componentWillMount() {
        this.setState({...this.state,
            listNameMaxHeight: window.innerHeight - 220 + "px",
            collapse: JSON.parse(localStorage.getItem('toolbox_collapse')) || false
        });
    }

    toggleCollapse = () => {
        let value = !this.state.collapse;
        localStorage.setItem('toolbox_collapse', value);
        this.setState({collapse: value});
    };

    render() {
        let img = window.__STATIC_URL__;
        let solved = this.props.regions.filter(obj => (obj.isSolved)).length;
        return <div className="toolbox_wrapper">
            <div className="toolbox">
                <div className="listname-wrapper">
                    {this.props.wsState !== true &&
                        <div id="network_connection_label">{localization.networkError}</div>}
                    {localization.found}: <span>{solved}</span>/<span>{this.props.regions.length}</span>
                    <span
                        className={"glyphicon glyphicon-chevron-" + (this.state.collapse ? 'up' : 'down')}
                        onClick={this.toggleCollapse}>
                    </span>
                    <Panel expanded={!this.state.collapse} onToggle={this.toggleCollapse}>
                        <Panel.Collapse>
                            <Panel.Body>
                                <div className="map_switcher_wrapper">
                                    <img className="map_switcher" src={img + "images/map/terrain.png"}
                                         onClick={() => {this.props.setMapType(google.maps.MapTypeId.TERRAIN)}}/>
                                    <img className="map_switcher" src={img + "images/map/hybrid.png"}
                                         onClick={() => {this.props.setMapType(google.maps.MapTypeId.HYBRID)}}/>
                                    <img className="map_switcher" src={img + "images/map/satellite.png"}
                                         onClick={() => {this.props.setMapType(google.maps.MapTypeId.SATELLITE)}}/>
                                </div>
                                <Scrollbars autoHide={true} autoHeight={true} autoHeightMax={this.state.listNameMaxHeight}>
                                    <ul className="list-group">
                                        {this.props.regions.map(polygon => (
                                            <NameListItem key={polygon.id} polygon={polygon}
                                                          click={() => this.props.openInfobox(polygon)}/>
                                        ))}
                                    </ul>
                                </Scrollbars>
                            </Panel.Body>
                        </Panel.Collapse>
                    </Panel>
                </div>
            </div>
        </div>;
    }
}


export default Toolbox;
