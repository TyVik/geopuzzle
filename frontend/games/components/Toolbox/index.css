.toolbox_wrapper {
    z-index: 3;
    width: 184px;
}

.toolbox {
    text-align: center;
    z-index: 2;
    background-color: white;
    border-radius: 5px;
    border: 1px solid black;
    max-width: 186px;
    display: flex;
    flex-direction: column;
}

.toolbox .panel {
    margin-bottom: 0;
}

.toolbox .panel-body {
    padding: 0;
}

.toolbox #network_connection_label {
    color: red;
    font-size: 10px;
}

.map_switcher_wrapper {
    display: flex;
    justify-content: center;
}

.map_switcher {
    width: 60px;
    height: 60px;
    cursor: pointer;
    padding: 3px;
}

.listname-wrapper {
    padding-top: 5px;
}

.listname-wrapper .glyphicon {
    float: right;
    cursor: pointer;
    padding-right: 5px;
}

.listname-wrapper .list-group {
    overflow-y: auto;
    overflow-x: hidden;
    margin-bottom: 0px;
}

.listname-wrapper .list-group-item {
    padding: 2px 4px;
    text-align: left;
}

.listname-wrapper .list-group-item.clickable {
    cursor: pointer;
}