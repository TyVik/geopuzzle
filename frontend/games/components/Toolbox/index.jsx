'use strict';
/* global google */
import React from "react";
import {Collapse} from "react-bootstrap";
import { Scrollbars } from 'react-custom-scrollbars';
import ListGroup from "react-bootstrap/ListGroup";
import {FormattedMessage as Msg} from "react-intl";

import "./index.css";


const NameListItem = (props) => {
  if (!props.polygon.draggable) {
    return <ListGroup.Item variant={props.polygon.isSolved ? 'success' : 'danger'} onClick={props.click} action={true}>
      {props.polygon.infobox.name}
    </ListGroup.Item>;
  } else {
    return <ListGroup.Item variant="danger">&nbsp;</ListGroup.Item>;
  }
};


class Toolbox extends React.Component {
  COLLAPSE_ID = 'toolbox-collapse';

  componentWillMount() {
    this.setState({...this.state,
      listNameMaxHeight: window.innerHeight - 220 + "px",
      collapse: JSON.parse(localStorage.getItem('toolbox_collapse')) || false
    });
  }

  toggleCollapse = () => {
    let value = !this.state.collapse;
    localStorage.setItem('toolbox_collapse', value);
    this.setState({collapse: value});
  };

  render() {
    let img = window.__STATIC_URL__;
    let solved = this.props.regions.filter(obj => (obj.isSolved)).length;
    return <div className="toolbox-wrapper">
      <div className="toolbox">
        <div className="toolbox-header" onClick={this.toggleCollapse} aria-controls={this.COLLAPSE_ID} aria-expanded={!this.state.collapse}>
          {this.props.wsState !== true &&
            <div id="network_connection_label"><Msg id="networkError"/></div>}
          <Msg id="found"/>: <span>{solved}</span>/<span>{this.props.regions.length}</span>
          <i className={"fas fa-angle-" + (this.state.collapse ? 'up' : 'down')} />
        </div>
        <Collapse in={!this.state.collapse}>
          <div id={this.COLLAPSE_ID}>
            <div className="map_switcher_wrapper">
              <img className="map_switcher" src={img + "images/map/terrain.png"}
                   onClick={() => {this.props.setMapType(google.maps.MapTypeId.TERRAIN)}}/>
              <img className="map_switcher" src={img + "images/map/hybrid.png"}
                   onClick={() => {this.props.setMapType(google.maps.MapTypeId.HYBRID)}}/>
              <img className="map_switcher" src={img + "images/map/satellite.png"}
                   onClick={() => {this.props.setMapType(google.maps.MapTypeId.SATELLITE)}}/>
            </div>
            <Scrollbars autoHide={true} autoHeight={true} autoHeightMax={this.state.listNameMaxHeight}>
              <ListGroup>
                {this.props.regions.map(polygon => (
                  <NameListItem key={polygon.id} polygon={polygon}
                                click={() => this.props.openInfobox(polygon)}/>
                ))}
              </ListGroup>
            </Scrollbars>
          </div>
        </Collapse>
      </div>
    </div>;
  }
}


export default Toolbox;
