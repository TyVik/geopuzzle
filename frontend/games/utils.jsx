import messages from "../locale/messages";

const prepareInfobox = (json) => {
  if (json.area) {
    json.area = `${Number(json.area).toLocaleString()} ${messages[window.__LANGUAGE__].km2}`;
  }
  if (json.population) {
    json.population = Number(json.population).toLocaleString();
  }
  ['depth', 'length', 'width', 'height'].forEach(item => {
    if (json[item]) {
      json[item] = `${Number(json[item]).toLocaleString()} ${messages[window.__LANGUAGE__].m}`;
    }
  });
  return json;
};

function shuffle(a) {
  for (let i = a.length; i; i--) {
    let j = Math.floor(Math.random() * i);
    [a[i - 1], a[j]] = [a[j], a[i - 1]];
  }
  return a;
}


export {prepareInfobox, shuffle};