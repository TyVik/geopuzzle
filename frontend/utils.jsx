'use strict';
import 'whatwg-fetch';
import Cookies from 'js-cookie';

import localization from "./localization";

function moveTo(paths, from, to) {
    let newPoints =[];
    let newPaths = [];

    for (let p = 0; p < paths.length; p++) {
        let path = paths[p];
        newPoints.push([]);

        for (let i = 0; i < path.length; i++) {
            newPoints[newPoints.length-1].push({
                heading: google.maps.geometry.spherical.computeHeading(from, path[i]),
                distance: google.maps.geometry.spherical.computeDistanceBetween(from, path[i])
            });
        }
    }

    for (let j = 0, jl = newPoints.length; j < jl; j++) {
        let shapeCoords = [],
            relativePoint = newPoints[j];
        for (let k = 0, kl = relativePoint.length; k < kl; k++) {
            shapeCoords.push(google.maps.geometry.spherical.computeOffset(
                to,
                relativePoint[k].distance,
                relativePoint[k].heading
            ));
        }
        newPaths.push(shapeCoords);
    }
    return newPaths;
}


function decodePolygon(polygon) {
    return polygon.map(polygon => (google.maps.geometry.encoding.decodePath(polygon)));
}

const prepareInfobox = (json) => {
    if (json.area) {
        json.area = Number(json.area).toLocaleString() + ' ' + localization.km2;
    }
    if (json.population) {
        json.population = Number(json.population).toLocaleString();
    }
    return json;
};

function shuffle(a) {
    for (let i = a.length; i; i--) {
        let j = Math.floor(Math.random() * i);
        [a[i - 1], a[j]] = [a[j], a[i - 1]];
    }
    return a;
}


const CSRFfetch = (url, options) => {
    let headers = options.headers || new Headers();
    headers.append('X-CSRFTOKEN', Cookies.get('csrftoken'));
    return fetch(url, {...options, headers: headers, credentials: 'same-origin'});
};


export {moveTo, decodePolygon, prepareInfobox, shuffle, CSRFfetch};
