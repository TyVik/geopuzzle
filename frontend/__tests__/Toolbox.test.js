'use strict';

import React from "react";
import {shallow} from "enzyme";
import Toolbox from "../games/components/Toolbox";
import {createComponentWithIntl, mountComponentWithIntl} from "./utils";


describe('shallow <Toolbox /> components', () => {
  let onSetMapType = jest.fn();
  let onOpenInfobox = jest.fn();
  let props = {setMapType: onSetMapType, regions: [], wsState: true, openInfobox: onOpenInfobox};

  it('check props', () => {
    let wrapper = shallow(<Toolbox {...props} regions={global.REGIONS}/>);
    expect(wrapper).toMatchSnapshot('toolbox');
  });

  it('check map type switcher', () => {
    let wrapper = shallow(<Toolbox {...props}/>);
    let switchers = wrapper.find('.map-switcher').getElements();
    expect(switchers.length).toBe(3);
    [google.maps.MapTypeId.TERRAIN, google.maps.MapTypeId.HYBRID, google.maps.MapTypeId.SATELLITE].map((item, index) => {
      switchers[index].props.onClick();
      expect(onSetMapType).toHaveBeenCalledWith(item);
    });
  });

  it('ws disconnect', () => {
    let wrapper = shallow(<Toolbox {...props}/>);
    expect(wrapper.find('#network-connection-label').length).toBe(0);

    wrapper = shallow(<Toolbox {...props} wsState={false}/>);
    expect(wrapper.find('#network-connection-label').length).toBe(1);
  });

  it('click collapse', () => {
    let wrapper = mountComponentWithIntl(<Toolbox {...props} regions={global.REGIONS}/>);
    let header = wrapper.find('.toolbox-header');
    expect(localStorage.getItem('toolbox-collapse')).toBe(null);
    expect(wrapper.find('#toolbox-collapse').hasClass('show')).toBeTruthy();
    expect(header.find('i').hasClass('fa-angle-down')).toBeTruthy();

    header.simulate('click');
    wrapper.update();
    expect(localStorage.getItem('toolbox-collapse')).toBe('true');
    expect(wrapper.find('#toolbox-collapse').hasClass('show')).toBeFalsy();
    expect(wrapper.find('i').hasClass('fa-angle-up')).toBeTruthy();
  });
});
