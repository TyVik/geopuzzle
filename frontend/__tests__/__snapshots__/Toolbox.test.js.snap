// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shallow <Toolbox /> components check props: toolbox 1`] = `
<div
  className="toolbox-wrapper"
>
  <div
    className="toolbox"
  >
    <div
      aria-controls="toolbox-collapse"
      aria-expanded={true}
      className="toolbox-header"
      onClick={[Function]}
    >
      <MemoizedFormattedMessage
        id="found"
      />
      : 
      <span>
        1
      </span>
      /
      <span>
        3
      </span>
      <i
        className="fas fa-angle-down"
      />
    </div>
    <ForwardRef
      appear={false}
      getDimensionValue={[Function]}
      in={true}
      mountOnEnter={false}
      timeout={300}
      unmountOnExit={false}
    >
      <div
        id="toolbox-collapse"
      >
        <div
          className="map-switcher-wrapper"
        >
          <img
            className="map-switcher"
            onClick={[Function]}
            src="/static/images/map/terrain.png"
          />
          <img
            className="map-switcher"
            onClick={[Function]}
            src="/static/images/map/hybrid.png"
          />
          <img
            className="map-switcher"
            onClick={[Function]}
            src="/static/images/map/satellite.png"
          />
        </div>
        <Scrollbars
          autoHeight={true}
          autoHeightMax="548px"
          autoHeightMin={0}
          autoHide={true}
          autoHideDuration={200}
          autoHideTimeout={1000}
          hideTracksWhenNotNeeded={false}
          renderThumbHorizontal={[Function]}
          renderThumbVertical={[Function]}
          renderTrackHorizontal={[Function]}
          renderTrackVertical={[Function]}
          renderView={[Function]}
          tagName="div"
          thumbMinSize={30}
          universal={false}
        >
          <ListGroup>
            <NameListItem
              click={[Function]}
              key="5054"
              polygon={
                Object {
                  "draggable": false,
                  "id": 5054,
                  "infobox": Object {
                    "area": "32,300",
                    "capital": Object {
                      "lat": 52,
                      "lng": 23,
                      "name": "Брест",
                      "wiki": "https://ru.wikipedia.org/wiki/Брест",
                    },
                    "coat_of_arms": "https://upload.wikimedia.org/wikipedia/commons/e/e7/Escut_Oblast_Brest.png",
                    "flag": "https://upload.wikimedia.org/wikipedia/commons/e/ec/Flag_of_Brest_Voblast%2C_Belarus.svg",
                  },
                  "isSolved": true,
                  "loaded": true,
                  "marker": Object {
                    "lat": 52,
                    "lng": 23,
                  },
                  "name": "Брестская область",
                  "paths": Array [],
                  "population": "1386982",
                  "wiki": "https://ru.wikipedia.org/wiki/Брестская_область",
                }
              }
            />
            <NameListItem
              click={[Function]}
              key="5060"
              polygon={
                Object {
                  "draggable": true,
                  "id": 5060,
                  "infobox": Object {
                    "loaded": false,
                    "name": "Витебская область",
                  },
                  "isSolved": false,
                  "paths": Array [],
                }
              }
            />
            <NameListItem
              click={[Function]}
              key="5055"
              polygon={
                Object {
                  "draggable": true,
                  "id": 5055,
                  "infobox": Object {
                    "loaded": false,
                    "name": "Гомельская область",
                  },
                  "isSolved": false,
                  "paths": Array [],
                }
              }
            />
          </ListGroup>
        </Scrollbars>
      </div>
    </ForwardRef>
  </div>
</div>
`;
